### PR TITLE
STORM-1038: Upgrade to Netty 4

### PR DESCRIPTION
**Pull request by srdo**
This is a continuation of the work done at https://github.com/apache/storm/pull/728.

### Important changes:
* Split MessageEncoder into multiple encoders depending on the message type, and made the stateless ones singletons
* Moved worker context termination to after the transfer thread has shut down. Shutting the context down first looks racy to me, since the transfer thread uses it.
* ChannelGroups automatically remove closed Channels, so Server/PacemakerServer doesn't bother removing closed channels manually anymore
* Made some changes to TransferDrainer for readability. I didn't see any performance impact from them, and they aren't necessary to upgrade to Netty 4. Let me know if they should go in another PR.

# Benchmarking
I benchmarked on a single-node install. Results are accessible at https://drive.google.com/open?id=1OuYHusyshQbzx38q5jchj8W98WG6An0L.

The branch point on master I compared to is fe32d95c61d0a4c9350eb4515f6e74b2aec761a3

## Speed of light
Ran `./storm jar .\storm-perf-2.0.0-SNAPSHOT.jar org.apache.storm.perf.ConstSpoutIdBoltNullBoltTopo 600 .\ConstSpoutIdBoltNullBoltTopo.yaml` with the following config

```
spout.count : 1
bolt1.count : 1  # IdBolt instances
bolt2.count : 1  # DevNullBolt instances

topology.workers : 3
```
This should force all messages to cross to another worker, which hopefully does a good job showing any decrease in throughput for the changed code.

## TVL
I didn't put much thought into setting the number of workers, or component counts here, let me know if I should try rerunning with some other numbers.

Ran `./storm jar .\storm-loadgen-2.0.0-SNAPSHOT.jar org.apache.storm.loadgen.ThroughputVsLatency --rate 90000 --spouts 4 --splitters 4 --counters 4 --reporter 'tsv:test.csv' -c topology.workers=4 --test-time 10`.

90k tuples is more than my machine can handle, so this is mostly to demonstrate that Storm doesn't choke any worse than it did before.

Also ran `./storm jar .\storm-loadgen-2.0.0-SNAPSHOT.jar org.apache.storm.loadgen.ThroughputVsLatency --rate 75000 --spouts 4 --splitters 4 --counters 4 --reporter 'tsv:test.csv' -c topology.workers=4 --test-time 10`

75k is right on the edge of what my computer can put through, so it shows what latencies look like compared to master. 

Performance looks pretty much the same to me pre- and post these changes. Let me know if there are other tests I should run to validate these changes.

# To do at some point
* ByteBufs are reference counted, so we should validate that we don't leak references when testing. Netty can tell us when a leak occurs, but I haven't set anything up to fail the build if these logs occur during testing yet https://netty.io/wiki/reference-counted-objects.html#leak-detection-levels. I did run some tests with the detection level set to paranoid, and didn't see any leaks.
* Didn't really test Pacemaker

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -82,7 +82,7 @@
             <roles>
                 <role>Committer</role>
             </roles>
-            <timezone/>
+            <timezone />
         </developer>
```
Comment (danny0405) [inline pom.xml]: nit whitespace

File: pom.xml [hunk 2]
```diff
@@ -109,7 +109,7 @@
...
```
Comment (danny0405) [inline pom.xml]: nit whitespace

**Comment by danny0405** [comment]
@srdo 
Great work, i will take my time to review this patch as soon as possible.

File: storm-client/src/jvm/org/apache/storm/messaging/netty/SaslStormClientHandler.java [hunk 3]
```diff
@@ -41,80 +38,88 @@ public SaslStormClientHandler(ISaslClient client) throws IOException {
...
-        SaslMessageToken saslTokenMessage = (SaslMessageToken) event
-            .getMessage();
+    }
+    
+    private void handleSaslMessageToken(ChannelHandlerContext ctx, SaslMessageToken saslMessageToken) throws Exception {
+        Channel channel = ctx.channel();
+        SaslNettyClient saslNettyClient = getChannelSaslNettyClient(channel);
         LOG.debug("Responding to server's token of length: "
-                  + saslTokenMessage.getSaslToken().length);
+                  + saslMessageToken.getSaslToken().length);
```
Comment (danny0405) [inline storm-client/src/jvm/org/apache/storm/messaging/netty/SaslStormClientHandler.java]: nit: indent

File: storm-client/src/jvm/org/apache/storm/messaging/netty/StormServerPipelineFactory.java [hunk 4]
```diff
@@ -12,28 +12,40 @@
...
 
-    public ChannelPipeline getPipeline() throws Exception {
+    @Override
+    protected void initChannel(Channel ch) throws Exception {
         // Create a default pipeline implementation.
-        ChannelPipeline pipeline = Channels.pipeline();
-
+        ChannelPipeline pipeline = ch.pipeline();
+        
         // Decoder
```
Comment (danny0405) [inline storm-client/src/jvm/org/apache/storm/messaging/netty/StormServerPipelineFactory.java]: nit: whitespace

File: storm-client/src/jvm/org/apache/storm/pacemaker/PacemakerClient.java [hunk 5]
```diff
@@ -100,23 +99,25 @@ public PacemakerClient(Map<String, Object> config, String host) {
...
+        this.workerEventLoopGroup = new NioEventLoopGroup(maxWorkers > 0 ? maxWorkers : 0, workerFactory);
+        int thriftMessageMaxSize = (Integer) config.get(Config.PACEMAKER_THRIFT_MESSAGE_SIZE_MAX);
+        bootstrap = new Bootstrap()
+            .group(workerEventLoopGroup)
+            .channel(NioSocketChannel.class)
+            .option(ChannelOption.TCP_NODELAY, true)
+            .option(ChannelOption.SO_SNDBUF, 5242880)
+            .option(ChannelOption.SO_KEEPALIVE, true)
+            .option(ChannelOption.WRITE_BUFFER_WATER_MARK, new WriteBufferWaterMark(8 * 1024, 32 * 1024))
+            .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
```
Comment (danny0405) [inline storm-client/src/jvm/org/apache/storm/pacemaker/PacemakerClient.java]: Can we just reuse this code block? Saw this in storm-client/src/jvm/org/apache/storm/messaging/netty/Server.java

File: storm-client/src/jvm/org/apache/storm/utils/TransferDrainer.java [hunk 6]
```diff
@@ -40,94 +41,39 @@ public void add(TaskMessage taskMsg) {
     }
 
     public void send(Map<Integer, NodeInfo> taskToNode, Map<NodeInfo, IConnection> connections) {
-        HashMap<NodeInfo, ArrayList<ArrayList<TaskMessage>>> bundleMapByDestination = groupBundleByDestination(taskToNode);
-
-        for (Map.Entry<NodeInfo, ArrayList<ArrayList<TaskMessage>>> entry : bundleMapByDestination.entrySet()) {
+        HashMap<NodeInfo, Stream<TaskMessage>> bundleMapByDestination = groupBundleByDestination(taskToNode);
+        
+        for (Map.Entry<NodeInfo, Stream<TaskMessage>> entry : bundleMapByDestination.entrySet()) {
```
Comment (danny0405) [inline storm-client/src/jvm/org/apache/storm/utils/TransferDrainer.java]: nit: whitespace

File: storm-client/src/jvm/org/apache/storm/utils/TransferDrainer.java [hunk 7]
```diff
@@ -40,94 +41,39 @@ public void add(TaskMessage taskMsg) {
...
         }
     }
 
-    private HashMap<NodeInfo, ArrayList<ArrayList<TaskMessage>>> groupBundleByDestination(Map<Integer, NodeInfo> taskToNode) {
-        HashMap<NodeInfo, ArrayList<ArrayList<TaskMessage>>> result = new HashMap<>();
-
+    private HashMap<NodeInfo, Stream<TaskMessage>> groupBundleByDestination(Map<Integer, NodeInfo> taskToNode) {
+        HashMap<NodeInfo, Stream<TaskMessage>> result = new HashMap<>();
+        
         for (Entry<Integer, ArrayList<TaskMessage>> entry : bundles.entrySet()) {
```
Comment (danny0405) [inline storm-client/src/jvm/org/apache/storm/utils/TransferDrainer.java]: nit: whitespace

File: storm-core/pom.xml [hunk 8]
```diff
@@ -140,7 +140,7 @@
             <artifactId>data.codec</artifactId>
             <scope>test</scope>
         </dependency>
-
+        
         <!--java-->
```
Comment (danny0405) [inline storm-core/pom.xml]: nit: whitespace

File: storm-core/pom.xml [hunk 9]
```diff
@@ -365,17 +365,17 @@
...
-             <filtering>true</filtering>
-         </resource>
+            </resource>
+            <resource>
+                <directory>${basedir}/src/resources</directory>
+                <includes>
+                    <include>storm-core-version-info.properties</include>
+                </includes>
+                <filtering>true</filtering>
+            </resource>
```
Comment (danny0405) [inline storm-core/pom.xml]: nits: whitespaces

File: storm-core/pom.xml [hunk 10]
```diff
@@ -456,29 +456,29 @@
...
+                                <includes>
+                                    <include>jvm/**/*.java</include>
+                                    <include>clj/**/*.clj</include>
+                                </includes>
+                            </source>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
```
Comment (danny0405) [inline storm-core/pom.xml]: Same as above

File: storm-server/src/main/java/org/apache/storm/pacemaker/PacemakerServer.java [hunk 11]
```diff
@@ -83,48 +88,51 @@ public PacemakerServer(IServerMessageHandler handler, Map<String, Object> config
...
+            .childOption(ChannelOption.SO_KEEPALIVE, true)
+            .childOption(ChannelOption.WRITE_BUFFER_WATER_MARK, new WriteBufferWaterMark(8 * 1024, 32 * 1024))
+            .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+            .childHandler(new ThriftNettyServerCodec(this, config, authMethod, thriftMessageMaxSize));
+
+        try {
+            ChannelFuture channelFuture = bootstrap.bind(new InetSocketAddress(port)).sync();
+            allChannels.add(channelFuture.channel());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
Comment (danny0405) [inline storm-server/src/main/java/org/apache/storm/pacemaker/PacemakerServer.java]: Code reuse as in storm-client/src/jvm/org/apache/storm/messaging/netty/Server.java

File: storm-client/src/jvm/org/apache/storm/pacemaker/PacemakerClient.java (hunk 5 above)
Comment (srdo) [reply storm-client/src/jvm/org/apache/storm/pacemaker/PacemakerClient.java]: It's not the same block. The code is different in Server

File: storm-server/src/main/java/org/apache/storm/pacemaker/PacemakerServer.java (hunk 11 above)
Comment (srdo) [reply storm-server/src/main/java/org/apache/storm/pacemaker/PacemakerServer.java]: I'm not sure what to replace this with. Do you have something in mind?

File: storm-core/pom.xml (hunk 9 above)
Comment (srdo) [reply storm-core/pom.xml]: This whitespace is because it was formatted incorrectly before

File: storm-core/pom.xml (hunk 10 above)
Comment (srdo) [reply storm-core/pom.xml]: This was also indented inconsistently (3 spaces where the rest of the file is 4)

File: storm-client/src/jvm/org/apache/storm/messaging/netty/Client.java [hunk 12]
```diff
@@ -135,19 +141,29 @@
...
+        this.eventLoopGroup = eventLoopGroup;
+        // Initiate connection to remote destination
+        bootstrap = new Bootstrap()
+            .group(this.eventLoopGroup)
+            .channel(NioSocketChannel.class)
+            .option(ChannelOption.TCP_NODELAY, true)
+            .option(ChannelOption.SO_SNDBUF, bufferSize)
+            .option(ChannelOption.SO_KEEPALIVE, true)
+            .option(ChannelOption.WRITE_BUFFER_WATER_MARK, new WriteBufferWaterMark(lowWatermark, highWatermark))
+            .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
```
[the remark this answers is not in the document]
Comment (srdo) [reply storm-client/src/jvm/org/apache/storm/messaging/netty/Client.java]: I went over all uses of ByteBuf and checked that they are either released or passed on to the next step of a pipeline (in which case Netty should release it).

I also ran some of the test topologies (e.g. the TVL topology) with `-Dio.netty.leakDetection.level=paranoid` in the worker childopts. Checked that it worked by intentionally not releasing a bytebuf. I didn't see any warnings about leaked buffers.

Does it matter whether the byte buffer is pooled or not for leak detection?

**Comment by revans2** [comment]
I ran some performance tests, but they were on my laptop with other things going on in the background so all I can really tell is that the 4.x code is withing noise for 3.x.  But the margin for noise is rather high because of background processes.

**Comment by srdo** [comment]
One of the tests is flaky, currently trying to figure out why. https://travis-ci.org/apache/storm/jobs/388836374. The client is reporting that it's connected before channelActive has been called on the server.

**Comment by srdo** [comment]
Rebased, and put in a work around for the flaky test. The test was using the fact that the client's channel to the server was open as indication that the channelConnected method must have finished executing on the server. That doesn't seem to be the case in Netty 4, the client channel may report that it is open while the server is still running channelActive.

File: shaded-deps/pom.xml [hunk 13]
```diff
@@ -251,6 +253,10 @@
                             <pattern>org.jboss.netty</pattern>
                             <shadedPattern>org.apache.storm.shade.org.jboss.netty</shadedPattern>
                         </relocation>
+                        <relocation>
```
Comment (revans2) [inline shaded-deps/pom.xml]: Is the org.jboss.netty rule needed still?

File: storm-client/pom.xml [hunk 14]
```diff
@@ -172,7 +172,7 @@
                 <!--Note - the version would be inherited-->
                 <configuration>
                     <excludes>**/generated/**</excludes>
-                    <maxAllowedViolations>3298</maxAllowedViolations>
+                    <maxAllowedViolations>9999</maxAllowedViolations>
```
Comment (revans2) [inline storm-client/pom.xml]: Can we set this to the correct number?

Comment (srdo) [reply storm-client/pom.xml]: Whoops, yes.

File: shaded-deps/pom.xml (hunk 13 above)
Comment (srdo) [reply shaded-deps/pom.xml]: Zookeeper depends on Netty 3. I think we still need to relocate it, even though our code doesn't use it directly.

Comment (revans2) [reply shaded-deps/pom.xml]: That's right I remember now

**Comment by HeartSaVioR** [comment]
Closed in 4aa6b5e18cf2dcefb509f965cd8a61e31389d6aa

**Pull request closed without merging**